### PR TITLE
Adds tests for database migration tool and refactors database migration executors

### PR DESCRIPTION
**Pull request by Boris-ILIEV**

### CODE REVIEW

File: com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/executor/DatabaseSequenceMigrationExecutorTest.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                                                                                      .build();
+    }
+
+    @Test
+    public void testExecuteMigrationInternalWhenLastSequenceValueIsZero() throws SQLException {
+        Mockito.when(mockSourceDatabaseQueryClient.getLastSequenceValue(TEST_SEQUENCE_NAME))
+               .thenReturn(ZERO_SEQUENCE_VALUE);
+
+        Assertions.assertThrows(IllegalStateException.class, () -> {
+            databaseSequenceMigrationExecutor.executeMigrationInternal(TEST_SEQUENCE_NAME);
```
Comment (radito3) [inline com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/executor/DatabaseSequenceMigrationExecutorTest.java:50]: Why the lambda body and not just the statement?
For this and all the following tests that use `assertThrows`

File: com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    public ResultSet getDataFromDataSource(String databaseTable) throws SQLException {
+        return executeWithAutoCommit((connection) -> {
+            PreparedStatement preparedStatement = null;
+            ResultSet resultSet = null;
+            try {
+                preparedStatement = prepareStatement(connection, String.format("SELECT * FROM %s", databaseTable));
+                resultSet = executeQuery(preparedStatement);
+                return resultSet;
+            } finally {
+                closeResultSetQuietly(resultSet);
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java:96]: Buut if you close the result set, what would this method return?

File: com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+public class DatabaseQueryClient {
+    private final SqlQueryExecutor sqlQueryExecutor;
+
+    public DatabaseQueryClient(SqlQueryExecutor sqlQueryExecutor) {
+        this.sqlQueryExecutor = sqlQueryExecutor;
+    }
+
+    public long getLastSequenceValue(String sequenceName) throws SQLException {
+        return executeWithAutoCommit((connection) -> {
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java:28]: You don't need the braces around the lambda's parameter if you have just one parameter.

File: com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            }
+        });
+    }
+
+    public void writeDataToDataSource(ResultSet resultSet, String insertQuery, DatabaseTableMetadata sourceTableMetadata)
+        throws SQLException {
+        execute((connection) -> {
+            try {
+                while (resultSet.next()) {
+                    transferDataToDatasource(connection, insertQuery, sourceTableMetadata, resultSet);
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java:107]: The `source` in `Datasource` is not capitalized here.

File: com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+    private PreparedStatement prepareStatement(Connection connection, String statement) throws SQLException {
+        return connection.prepareStatement(statement);
+    }
+
+    private ResultSet executeQuery(PreparedStatement preparedStatement) throws SQLException {
+        return preparedStatement.executeQuery();
+    }
+
+    private void closeStatementQuietly(PreparedStatement preparedStatement) {
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java:163]: Why not use `JdbcUtil.closeQuietly` directly?

File: com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        JdbcUtil.closeQuietly(preparedStatement);
+    }
+
+    private void closeResultSetQuietly(ResultSet resultSet) {
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java:167]: Why not use `JdbcUtil.closeQuietly` directly?

File: com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+    private <T> T executeWithAutoCommit(SqlQuery<T> sqlQuery) throws SQLException {
+        return sqlQueryExecutor.executeWithAutoCommit(sqlQuery);
+    }
+
+    private <T> T execute(SqlQuery<T> sqlQuery) throws SQLException {
+        return sqlQueryExecutor.execute(sqlQuery);
+    }
+
+    private PreparedStatement prepareStatement(Connection connection, String statement) throws SQLException {
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java:155]: What's the purpose of this method?

File: com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        return connection.prepareStatement(statement);
+    }
+
+    private ResultSet executeQuery(PreparedStatement preparedStatement) throws SQLException {
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/main/java/com/sap/cloud/lm/sl/cf/database/migration/client/DatabaseQueryClient.java:159]: What's the purpose of this method?

File: com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/executor/DatabaseTableMigrationExecutorTest.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        Mockito.verify(mockSourceDatabaseQueryClient)
+               .extractTableMetadataFromDatabase(TEST_DATABASE_TABLE);
+        Mockito.verify(mockSourceDatabaseQueryClient)
+               .getDataFromDataSource(TEST_DATABASE_TABLE);
+        Mockito.verify(mockTargetDatabaseQueryClient)
+               .writeDataToDataSource(RESULT_SET, "", DATABASE_TABLE_METADATA);
+    }
+
+    @Test
+    public void testExecuteMigrationInternalWithNullDatabaseTableString() throws SQLException {
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/executor/DatabaseTableMigrationExecutorTest.java:70]: You don't need to test what happens when the user supplies `null` or an empty string here.

File: com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/executor/type/DatabaseTypeSetterFactoryTest.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+class DatabaseTypeSetterFactoryTest {
+    private static final String EXISTING_DATABASE_TYPE_SETTER = "bool";
+
+    private static final String NON_EXISTING_DATABASE_TYPE_SETTER = "test";
+
+    private DatabaseTypeSetterFactory databaseTypeSetterFactory;
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/executor/type/DatabaseTypeSetterFactoryTest.java:14]: You don't need to have this field. Fields are used when you want to share state between methods, but you're changing the value in each method, so turn this into a local variable.

File: com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/generator/DatabaseTableInsertQueryGeneratorTest.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+import com.sap.cloud.lm.sl.cf.database.migration.metadata.DatabaseTableMetadata;
+import com.sap.cloud.lm.sl.cf.database.migration.metadata.ImmutableDatabaseTableColumnMetadata;
+import com.sap.cloud.lm.sl.cf.database.migration.metadata.ImmutableDatabaseTableMetadata;
+
+class DatabaseTableInsertQueryGeneratorTest {
+    private static final String TEST_TABLE_NAME = "testTableName";
+    private static final String TEST_COLUMN_NAME = "testColumnName";
+    private static final String TEST_COLUMN_TYPE = "testColumnType";
+
+    private DatabaseTableInsertQueryGenerator databaseTableInsertQueryGenerator;
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/generator/DatabaseTableInsertQueryGeneratorTest.java:19]: Just initialize the field here, so that you don't have to initialize it in every test.

File: com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/generator/DatabaseTableInsertQueryGeneratorTest.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+                                                                                                                                   TEST_COLUMN_TYPE));
+        DatabaseTableMetadata databaseTableMetadata = buildTestDatabaseTableMetadata(multipleDatabaseTableColumnMetadata, TEST_TABLE_NAME);
+
+        String expectedQuery = "INSERT INTO testTableName(testColumnName, testColumnName) VALUES (?, ?)";
+        String resultQuery = databaseTableInsertQueryGenerator.generate(databaseTableMetadata);
+
+        Assertions.assertEquals(expectedQuery, resultQuery);
+    }
+
+    private ImmutableDatabaseTableColumnMetadata buildTestDatabaseTableColumnMetadata(String columnName, String columnType) {
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/generator/DatabaseTableInsertQueryGeneratorTest.java:49]: Rename to `buildDatabaseTableColumnMetadata`. Yes, you're using the result for testing purposes, but this method doesn't need to know that.

File: com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/generator/DatabaseTableInsertQueryGeneratorTest.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        return ImmutableDatabaseTableColumnMetadata.builder()
+                                                   .columnName(columnName)
+                                                   .columnType(columnType)
+                                                   .build();
+    }
+
+    private ImmutableDatabaseTableMetadata
+            buildTestDatabaseTableMetadata(List<DatabaseTableColumnMetadata> multipleDatabaseTableColumnMetadata, String tableName) {
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/generator/DatabaseTableInsertQueryGeneratorTest.java:57]: Rename to `buildDatabaseTableMetadata`. Yes, you're using the result for testing purposes, but this method doesn't need to know that.

File: com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/generator/DatabaseTableInsertQueryGeneratorTest.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    private DatabaseTableInsertQueryGenerator databaseTableInsertQueryGenerator;
+
+    @Test
+    public void testGenerateWithTestDatabaseMetadataWhenSingleColumn() {
+        databaseTableInsertQueryGenerator = new DatabaseTableInsertQueryGenerator();
+        DatabaseTableColumnMetadata databaseTableColumnMetadata = buildTestDatabaseTableColumnMetadata(TEST_COLUMN_NAME, TEST_COLUMN_TYPE);
+        DatabaseTableMetadata databaseTableMetadata = buildTestDatabaseTableMetadata(Arrays.asList(databaseTableColumnMetadata),
+                                                                                     TEST_TABLE_NAME);
+
+        String expectedQuery = "INSERT INTO testTableName(testColumnName) VALUES (?)";
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/generator/DatabaseTableInsertQueryGeneratorTest.java:28]: It's generally a good idea to define variables exactly before you actually need them. You don't need this variable in the next line, so move this declaration one line below.

File: com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/generator/DatabaseTableInsertQueryGeneratorTest.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        String resultQuery = databaseTableInsertQueryGenerator.generate(databaseTableMetadata);
+
+        Assertions.assertEquals(expectedQuery, resultQuery);
+    }
+
+    @Test
+    public void testGenerateWithTestDatabaseMetadataWhenMultipleColumns() {
+        databaseTableInsertQueryGenerator = new DatabaseTableInsertQueryGenerator();
+        List<DatabaseTableColumnMetadata> multipleDatabaseTableColumnMetadata = Arrays.asList(buildTestDatabaseTableColumnMetadata(TEST_COLUMN_NAME,
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.database/src/test/java/com/sap/cloud/lm/sl/cf/database/migration/generator/DatabaseTableInsertQueryGeneratorTest.java:37]: Put different names for each column.

**Comment by nictas** [comment]
Continued in: https://github.com/cloudfoundry-incubator/multiapps-controller/pull/901
